Blog; Using Snapshots for Version Control in Node-RED with FlowFuse (#2613)
FlowFuse/website: src/blog/2024/09/node-red-version-control-with-snapshots.md

## Managing Snapshots in FlowFuse

Creating snapshots in FlowFuse is straightforward and can be done in just a few steps.

### Creating Snapshots for Cloud and Device Instances

With this understanding, let’s walk through the steps to create a snapshot in FlowFuse. But before that, let’s understand the differences between a cloud instance and a device instance.

- **Cloud Instance**: This is a Node-RED instance running on a server's virtual machine.

- **Device Instance**: This refers to a Node-RED instance running on your edge devices, which are connected to the FlowFuse platform through the [FlowFuse Device Agent](/product/device-agent/).

Taking snapshots for cloud instances and device instances is completely different. When a device is not assigned to an instance, it will have the same interface and the snapshot tab for creating snapshots, allowing you to create the snapshot in a similar way as you would for an cloud instance. However, once you assign any device to the cloud instance, you will need to follow a different procedure to create device snapshots manually.

> #### Creating Snapshots for Cloud Instance

Reviewer: since we state above that the experience is almost the same for application devices and cloud instances, should this title and the below content not reflect that?

I mention this because there are 2 major sections in this article. 1. "Creating Snapshots for Cloud Instance", 2. "Creating Snapshot for Device Instances when assigned to Cloud Instance".  i.e. there is no _3. Creating Application Device snapshots_

Once a snapshot is created, it will be visible in a list format. Each snapshot will display details such as the name, description, creator, and creation date, along with a three-dot icon. 

![Image showing available options for managing snapshots in FlowFuse.](./images/options-available-for-instance.png){data-zoomable}
*Image showing available options for managing snapshots in FlowFuse.*

Clicking on the icon will open different options to manage and operate the snapshot.

### Creating Snapshot for Device Instances when assigned to Cloud Instance

To create a snapshot for the device instances when that device is assinged to cloud instance:

1. Go to **Devices** by clicking on the **Devices** option in the sidebar, and then click on the specific device.
2. Once clicked, a similar interface will open as with instances, allowing you to manage and monitor its settings and configuration. Ensure that the device has developer mode enabled.

![Image showing the 'Create Snapshot' option in the Developer mode tab](./images/option-to-create-device-snapshot.png
){data-zoomable}
*Image showing the 'Create Snapshot' option in the Developer mode tab*

> 3. Switch to the **Developer Mode** tab by clicking on the **Developer Mode** option at the top. In the Developer Mode section, you will see the **Create Snapshot** button. Click on it to create a snapshot, then enter the details such as the name and description. If needed, set it as the device target snapshot, and then click **Create**.

Reviewer: The way this reads to me is:

1. I must go to the instance -> devices view
2. Switch it into developer mode
3. Create a snapshot.

This is both counter productive and a little misleading.

If a device is in fleet mode, it WILL be running the "active snapshot" specified on the devices overview. 

I suspect we need to somehow make it clear that this is really only necessary if the device is in developer mode as it may have been modified. Perhaps something as simple as stating "typically, the device will be in fleet mode and will be running the flows specified by the target snapshot, however, if the device is in developer mode it may have been modified, in which case, you can take a snapshot directly from the device using the developer tab - here's how..."

   - **Prev**: Navigate to the previous difference between the two snapshots.
   - **Next**: Navigate to the next difference.

This comparison feature helps you easily identify changes, making it simple to spot issues or track flow modifications between versions.

### Downloading, Uploading, and Deploying Snapshots

Created snapshots can be downloaded locally, providing you with a backup of your Node-RED flows. This backup can be uploaded to another instance or restored in case of emergencies.

#### Downloading Snapshot:

![Image showing the 'Download Snapshot' option in the menu](./images/download-snapshot.png){data-zoomable}
*Image showing the 'Download Snapshot' option in the menu*

1. Click on the **three-dot** icon next to the corresponding snapshot, then select **Download Snapshot** from the menu.

![Image showing the prompt to download a snapshot in FlowFuse, featuring a highlighted key and the download button.](./images/download-snapshot-prompt.png){data-zoomable}
*Image showing the prompt to download a snapshot in FlowFuse, featuring a highlighted key and the download button*

> 2. A new prompt will open, displaying a **secret key**. This key is used to encrypt the configuration details of the snapshot. Copy the key and store it safely, as you will need it for restoring the snapshot.
>    - You can also generate a new secret key by clicking the **Random Secret** button on the right.

Reviewer: ```suggestion
2. A **Download Snapshot** dialog will open, displaying a **secret key**. This is used to encrypt sensitive info in the snapshot. As a convenience, a random secret is auto generated however you should change this to something memorable as it will be needed when you need to upload the snapshot.
```

Suggestion, less techy, More friendly?

3. After copying the secret key, click the **Download** button located at the bottom right of the prompt to download the snapshot.

#### Uploading Snapshots:

![Image showing the button to upload a snapshot in FlowFuse.](./images/button-upload-snapshot.png){data-zoomable}

> *Image showing the button to upload a snapshot in FlowFuse.*

Reviewer: Should retake this screenshot with a file selected that actually has credentials - since it will show the "secret2 field (this screenshot does not contain the secret field)

ALSO NOTE:

Not all snapshots have credentials. They are only present when a flows node has a secret entered (like an MQTT Broker config node).  So when a snapshots flows do NOT have creds, the secret will NOT be asked and will not be shown on the download OR upload.

1. Go to the **Snapshots** tab and click on the three-dot icon next to the desired snapshot.

![Image showing the 'Set as Device Target' option in the menu](./images/set-targe-snapshot-option.png){data-zoomable}
*Image showing the 'Set as Device Target' option in the menu*

2. Select **Set as Device Target** from the menu, prompt will open to conform then click on to the "Set Target".

![Image showing the green status indicating how many devices have deployed this snapshot option in the menu.](./images/status-showing-the-target-snapshot.png){data-zoomable}
*Image showing the green status indicating how many devices have deployed this snapshot option in the menu.*

Once done you will be able to see the green mark in that snapshot showing on how much devices it is deployed on.

## Common Mistakes to Avoid

1. **Neglecting Regular Backups**: Don’t skip those snapshots! Regularly create backups, especially before making significant changes, moving devices through instances or applications, or disabling developer mode. Think of it as your safety net—ensuring you can always bounce back to a stable state if unexpected issues pop up.

> 2. **Overlooking Auto-Snapshot Limits**: Did you know that your **auto snapshots** have limits? Be mindful of how many you can retain, as older **auto snapshots** will be automatically deleted. If you have important **auto snapshots**, save them locally to keep them safe. Remember, manually created snapshots have no limits, so take advantage of that!

Reviewer: ```suggestion
2. **Overlooking Auto-Snapshot Limits**: Did you know that your **auto snapshots** have limits? Be mindful of how many you can retain, as older **auto snapshots** will be automatically deleted. If you have important **auto snapshots**, either rename them to avoid automatic deletion or download and save them locally to keep them safe. Remember, manually created snapshots have no limits, so take advantage of that!
```

3. **Inadequate Secret Key Security**: Your secret key is your golden ticket! When downloading snapshots locally, **securely store that key**. Losing it could mean losing access to your snapshot and your ability to recover your work. Treat it like a password—keep it safe!

> 4. **Taking Device Snapshots in the Instance Snapshots Tab**: Make sure you're in the right place! Always take device snapshots in the **Developer Mode** tab. Some users mistakenly capture snapshots for devices in the instance associated with the devices, which only captures the snapshot for the instance itself. This can lead to losing valuable work done on specific device. 

Reviewer: Slightly misleading. As before, the instance device, when in fleet mode, should be running the snapshot set as "target" (will actually be running "active snapshot").

It is worth noting that if the device is in developer mode it MAY have modified flows and therefore using the correct snapshot action for an instance device becomes important.

---
title: "Using Snapshots for Version Control in Node-RED with FlowFuse"
subtitle: "Effortlessly manage and recover your Node-RED flows with snapshots in FlowFuse."
description: Learn how to use snapshots for version control in Node-RED with FlowFuse. Explore step-by-step guidance on creating, managing, and restoring flow backups to enhance your Node-RED development and save yourself from accidental changes.
date: 2024-09-25
authors: ["sumit-shinde", "stephen-mclaughlin"]
image: 
tags:
   - posts
   - node red version control
   - nodered version control
   - node-red flow backup
---

> Version control is essential, especially when multiple people are working on the same Node-RED project. Without it, changes can easily overlap, or worse—accidental updates could break critical flows. FlowFuse solves this challenge with snapshots, allowing you to create backups of your flows, restore previous versions, and safeguard your project from unexpected issues. 

Reviewer 2: ```suggestion
Version control is essential, especially when multiple people are working on the same Node-RED project. Without it, changes can easily overlap, or worse — accidental updates could break critical flows. FlowFuse solves this challenge with snapshots, allowing you to create backups of your flows, restore previous versions, and safeguard your project from unexpected issues. 
```

<!--more-->

> In this guide, we’ll show you how to use snapshots in FlowFuse to manage your Node-RED projects with confidence and prevent costly mistakes.

Reviewer 2: ```suggestion
Let's look at how to use snapshots in FlowFuse to manage your Node-RED projects with confidence and prevent costly mistakes.
```

   - **Prev**: Navigate to the previous difference between the two snapshots.
   - **Next**: Navigate to the next difference.

This comparison feature helps you easily identify changes, making it simple to spot issues or track flow modifications between versions.

### Downloading, Uploading, and Deploying Snapshots

Created snapshots can be downloaded locally, providing you with a backup of your Node-RED flows. This backup can be uploaded to another instance or restored in case of emergencies.

#### Downloading Snapshot:

![Image showing the 'Download Snapshot' option in the menu](./images/download-snapshot.png){data-zoomable}
*Image showing the 'Download Snapshot' option in the menu*

1. Click on the **three-dot** icon next to the corresponding snapshot, then select **Download Snapshot** from the menu.

![Image showing the prompt to download a snapshot in FlowFuse, featuring a highlighted key and the download button.](./images/download-snapshot-prompt.png){data-zoomable}
*Image showing the prompt to download a snapshot in FlowFuse, featuring a highlighted key and the download button*

> 2. A **Download Snapshot** dialog will open.  Note: If the flows contain sensitive values, an additional field for **secret key** will be show. This is used to encrypt the sensitive values in the snapshot. As a convenience, a random secret is auto generated however you should change this to something memorable as it will be needed when you later upload the snapshot.

Reviewer: ```suggestion
2. A **Download Snapshot** dialog will open.  Note: If the flows contain sensitive values, an additional field for **secret key** will be displayed. This is used to encrypt the sensitive values in the snapshot. As a convenience, a random secret is auto generated however you should change this to something memorable as it will be needed when you later upload the snapshot.
```

> 2. A **Download Snapshot** dialog will open, displaying a **secret key** if your snapshot flow has config nodes otherwise it not be displayed copy it if it is avialbe. This is used to encrypt sensitive info in the snapshot. As a convenience, a random secret is auto generated however you should change this to something memorable as it will be needed when you need to upload the snapshot.

Reviewer: ```suggestion
2. A **Download Snapshot** dialog will open.  Note: If the flows contain sensitive values, an additional field for **secret key** will be show. This is used to encrypt the sensitive values in the snapshot. As a convenience, a random secret is auto generated however you should change this to something memorable as it will be needed when you later upload the snapshot.
```

![Image showing the green status indicating how many devices have deployed this snapshot option in the menu.](./images/status-showing-the-target-snapshot.png){data-zoomable}
*Image showing the green status indicating how many devices have deployed this snapshot option in the menu.*

Once done you will be able to see the green mark in that snapshot showing on how much devices it is deployed on.

## Common Mistakes to Avoid

1. **Neglecting Regular Backups**: Don’t skip those snapshots! Regularly create backups, especially before making significant changes, moving devices through instances or applications, or disabling developer mode. Think of it as your safety net—ensuring you can always bounce back to a stable state if unexpected issues pop up.

2. **Overlooking Auto-Snapshot Limits**: Did you know that your **auto snapshots** have limits? Be mindful of how many you can retain, as older **auto snapshots** will be automatically deleted. If you have important **auto snapshots**, either rename them to avoid automatic deletion or download and save them locally to keep them safe. Remember, manually created snapshots have no limits, so take advantage of that!

> 3. **Inadequate Secret Key Security**: Your secret key is crucial for accessing snapshots that contain config nodes. When downloading snapshots locally, securely store that key. Losing it could mean losing access to your snapshot and your ability to recover your work. Treat it like a password—keep it safe!

Reviewer: ```suggestion
3. **Secret Key**: The secret key used to encrypt your snapshot is crucial for when you later need to upload that snapshot. When downloading snapshots locally, securely store that key. Losing it could mean losing access to your snapshot and your ability to recover your work. Treat it like a password—keep it safe!
```

> 4. **Taking Device Snapshots in the Instance Snapshots Tab**: Sometimes, device flows may have been modified while in **developer mode.** In that case, always take device snapshots in the **Developer Mode** tab. Some users mistakenly capture snapshots for devices in the instance associated with the devices, which only captures the snapshot for the instance itself and does not include the updates made on the device itself. This can lead to losing valuable work done on specific devices.

Reviewer: ```suggestion
4. **Taking a Snapshots of the wrong thing**: When a device is owned by an instance it typically runs the same flows however, if a device flows may have been modified directly in **developer mode** it will have different flows to the instance that owns it. In this case, before you move the device or switch it out of developer mode, it is recommended that you take a snapshot directly from the device itself. Direct device snapshots are performed on the **Developer Mode** tab of the device.
```

Suggestion - a bit less repetitive? A bit more of the "why".  see what you think
